python/pyfury/lib/meta/meta_string.py

```diff
@@ -0,0 +1,365 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import numpy as np
+from enum import Enum
+
+
+class Encoding(Enum):
+    UTF_8 = 0x00
+    LOWER_SPECIAL = 0x01
+    LOWER_UPPER_DIGIT_SPECIAL = 0x02
+    FIRST_TO_LOWER_SPECIAL = 0x03
+    ALL_TO_LOWER_SPECIAL = 0x04
+
+
+class MetaString:
+    def __init__(
+        self, original, encoding, special_char1, special_char2, encoded_data, length
+    ):
+        self.original = original
+        self.encoding = encoding
+        self.special_char1 = special_char1
+        self.special_char2 = special_char2
+        self.encoded_data = encoded_data
+        self.length = length
+        if self.encoding != Encoding.UTF_8:
+            self.strip_last_char = (encoded_data[0] & 0x80) != 0
+        else:
+            self.strip_last_char = False
+
+
+class MetaStringEncoder:
+    def __init__(self, special_char1, special_char2):
+        self.special_char1 = special_char1
+        self.special_char2 = special_char2
+
+    def encode(self, input_string):
+        if not input_string:
+            return MetaString(
+                input_string,
+                Encoding.UTF_8,
+                self.special_char1,
+                self.special_char2,
+                np.array([], np.uint8),
+                0,
+            )
+
+        encoding = self.compute_encoding(input_string)
+        return self.encode_with_encoding(input_string, encoding)
+
+    def encode_with_encoding(self, input_string, encoding):
+        if not input_string:
+            return MetaString(
+                input_string,
+                Encoding.UTF_8,
+                self.special_char1,
+                self.special_char2,
+                np.array([], np.uint8),
+                0,
+            )
+
```

Comment:
Maybe we should check the length of `input_string` and limit it to `Short.MAX_VALUE`.